🐛 Check Content-Type request header before assuming JSON

patrickkwang opened the pull request:
🐛 Check Content-Type request header before assuming JSON

Related to #1018.

---

**Edit** by @tiangolo: Updated tests to include extra corner cases, and updated implementation to parse content-type headers using Python's standard library `email` module to cover more corner cases.

Commit:
Check Content-Type before assuming JSON

Don't try to convert to JSON unless Content-Type == application/json.

Commit:
Remove now-unneccessary exception handling

Commit:
Fix broken-body test

ArcLightSlavik commented on fastapi/routing.py:
```diff
@@ -155,16 +154,20 @@ def get_request_handler(
-        except json.JSONDecodeError as e:
-            raise RequestValidationError([ErrorWrapper(e, ("body", e.pos))], body=e.doc)
+                        if (
+                            request.headers.get("Content-Type", "application/json")
```
Is Content-Type not being there a concern? I believe RFC says it should always be filled.

patrickkwang replied:
[RFC 7231](https://tools.ietf.org/html/rfc7231#section-3.1.1.5) suggests it is not strictly required, and that we should assume "application/octet-stream" if it's not present.

ArcLightSlavik commented on tests/test_tutorial/test_body/test_tutorial001.py:
```diff
@@ -178,16 +178,9 @@ def test_post_broken_body():
-                "msg": "Expecting value: line 1 column 1 (char 0)",
-                "type": "value_error.jsondecode",
+                "loc": ["body"],
+                "msg": "value is not a valid dict",
```
Would be considered a breaking change. Think even my work API has a test similar to this.

Commit:
Assume "Content-Type": "application/octet-stream"

Per RFC 7231, the "Content-Type" header is not strictly required, and if it is not present, "application/octet-stream" should be assumed.

Commit:
Explicitly specify Content-Type in gzip test

ArcLightSlavik reviewed (approved):
Find the formatting a bit odd but if that's how flake8(?) did it then okay.

Mause commented:
I think this needs to be thought about a bit more - there are other mime types that are JSON but which don't use the application/json minetype. application/hal+json is one such common example of this

patrickkwang commented:
Good point, @Mause. Based on [this list](https://www.iana.org/assignments/media-types/media-types.xhtml) of registered mimetypes, it seems safe to interpret as JSON any mimetype that _ends_ with "json". Notably, this excludes the "json-seq" type, which is not itself a subset of JSON.

Commit:
Interpret as JSON any mimetype ending with "json"

Mause commented:
Another thing - you're also breaking content-type headers with charset or boundary values, such as `application/json; charset=UTF-8`

Commit:
Generalize media type parsing to handle parameters

patrickkwang commented:
Perhaps some new tests are called-for at this point.

patrickkwang committed:
Add test for handling various Content-Type headers

patrickkwang commented:
How do we like conflicts to be handled? I can rebase and force-push to my feature branch...

rdpravin1895 commented:
Hi @patrickkwang. Are you planning to merge this?

patrickkwang commented:
I've rebased and everything is green, but I do not have the necessary permissions to merge.

rdpravin1895 commented:
Hi @tiangolo, sorry to disturb you, but can you have a look at this? We've been using FastAPI, so far has been an excellent framework, but only because of strict json application type, we are facing troubles in documenting the examples.

tiangolo committed:
✅ Update tests for content-type headers and include extra corner case

tiangolo committed:
♻️ Re-implement JSON content-type parsing with Python's standard library email module

to cover corner cases

tiangolo committed:
🔀 Merge master to update codebase, dependencies and tests

tiangolo committed:
✅ Include again test with patch to force exception, for coverage

tiangolo commented:
Thanks @patrickkwang ! :rocket: :nerd_face: 

And thanks for the thorough reviews @ArcLightSlavik and @Mause :bow: :cake: 

I updated the tests to include extra corner cases and updated the implementation to parse the content-type headers using Python's standard library `email` module to cover more corner cases.

jolynch commented:
For what it's worth this appears to be a pretty big breaking change for clients who were previously not sending `Content-Type` headers. For example, some of our production clients started receiving a validation error from a route where previously it was working fine:
```
{"message":"Error processing request","error":"RequestValidationError","details":[{"loc":["body"],"msg":"value is not a valid dict","type":"type_error.dict"}]}
```

I understand the original default was incorrect but this fix breaks functional code and I believe the CVE would be resolved by assuming json in any case that is not exempted from CORS preflight (so don't assume json when the content-type is explicitly set to something not `application/json`)? Alternatively could the server author configure the default assumed content type?

kpflugshaupt commented:
This just cost me two days of debugging. Not so pleased right now.

I fully see and appreciate the reasoning behind stricter checks, mind you. Go ahead with that!
But the error message is rather misleading.  _'value is not a valid dict'_ does **not** make me suspect my HTTP headers!

Is it possible to implement an exception that's more to the point here? I have the feeling I'm not the only one tearing their hair out over this...

Sorry for the venting -- FastAPI is and remains my favourite framework. Just feeling a bit betrayed by that error message?

hademircii commented:
> For what it's worth this appears to be a pretty big breaking change for clients who were previously not sending Content-Type headers. 
experiencing a similar problem right now (content-type is plain text). Bumping Fast API to 0.65.2 to breaks their integration, and we are not able to update all clients right away. So we either need to hold off on this and future FastAPI updates, or find a server side fix to go around it.

tiangolo commented:
That's a very good point @jolynch! Thanks for idea.

I just released version `0.65.3` that does exactly that. If there's no Content-Type header, it is now assumed as JSON.

This still provides protection from CSRF because browsers automatically set Content-Type headers. And if an attacker removes the header explicitly, it triggers the CORS preflight, which provides the main protection using CORS.

So, you should all be able to upgrade to `0.65.3` more easily now. :tada:
